### ISSUE TITLE
Function pointer in Hash literal crashes compiler

### ISSUE THREAD

**Issue by rhysd** [manastech/crystal#646]
- Code

```ruby
module Foo
  def self.blah
  end

  {"a" => ->self.blah}
end
```

- Error message

```
Nil assertion failed
*raise<String>:NoReturn +70 [4404349830]
*Nil#not_nil!<Nil>:NoReturn +13 [4404421037]
*Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::FixEmptyTypesVisitor>:Nil +2635 [4407051627]
*Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::FixEmptyTypesVisitor>:Nil +2423 [4407051415]
*Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::FixEmptyTypesVisitor>:Nil +2218 [4407051210]
*Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::ASTNode+, Crystal::FixEmptyTypesVisitor>:Nil +892 [4407049884]
*Crystal::Program#fix_empty_types<Crystal::Program, Crystal::Expressions>:Nil +275 [4407048963]
*Crystal::Compiler#compile<Crystal::Compiler, Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result +4434 [4406462162]
*Crystal::Command::run_command<Array(String)>:Nil +388 [4404394676]
*Crystal::Command::run<Array(String)>:(Nil | CFileIO | File | Bool | Crystal::Compiler::Result | Array(Crystal::Init::View+:Class)) +13832 [4404382456]
main +15985 [4404366545]

Error: you've found a bug in the Crystal compiler. Please open an issue: https://github.com/manastech/crystal/issues
```

- Environment
  - Crystal 0.7.1
  - OS X 10.9

**Comment by rhysd** [comment]
I could reproduce the bug outside module.

```ruby
def blah
end
b = {"a" => ->blah}
```

**Comment by asterite** [comment]
Thanks for reporting this, @rhysd . Luckily the fix was pretty simple :-)

**Comment by rhysd** [comment]
@asterite 

Thank you for quick fix!
I also confirmed the fix in fff0654.
